### PR TITLE
CLI: Automatically detect typescript in `sb init`

### PR DESCRIPTION
**Pull request by yannbf**
Issue: #5050 

## What I did
This PR adds automatic detection for typescript projects when installing storybook by running `npx @storybook/cli sb init`.
With that, it's not necessary to pass `--story-format` anymore (unless it is desired)!

While I was at it, I revamped the framework detection code to be more scalable. If you need to add a new template, all you have to do is create a new entry at `supportedTemplates` in `project_types.js` like the following:
```js
{
    preset: projectTypes.NEW_SUPPORTED_TEMPLATE,
    dependencies: [], // <-- optional, tests for these both as dependencies and devDependencies
    peerDependencies: [], // <-- optional
    files: [], // <-- optional
    matcherFunction: ({ dependencies, files, peerDependencies }) => {
      // every argument is returned as an array of booleans.
      return // whatever assertion you want, as long as it returns boolean.
    },
}
```
As for the typescript detection heuristics, I believe the common point in typescript projects is `typescript` as dependency. We can definitely change it to use more assertions, such as checking if files like `tsconfig.json` or `tslint.json` exist, or if the project contains at least one `*.ts|*.tsx` file.

I tried really hard to test and assert that the implementation works well, and comparing with the current implementation (in next), this is the only difference in the output (before/after):

But of course there could be edge cases or something I missed, so please help me out make sure this is good. Also please be thorough in the review, spare no comments! even with the simplest thing. 👍 In case you want to discuss, we can have a call if that makes it easier!

## How to test
Checkout this branch, run `yarn build cli`, go to `libs/cli` and run `yarn test`. Tweak with the test script if you want/ You can also check and run the unit tests `helpers.test.js` and `detect.test.js`.

## Does this need an update to the documentation?
I guess [this page](https://storybook.js.org/docs/configurations/typescript-config/) could be updated so the init command is mentioned there, however the page is about general typescript configuration, and currently storybook doesn't have templates for every framework + ts.

## Next steps
* Turn the files to typescript, so there's no way to make silly mistakes/typo and so that the code is easier to understand and use. I added comments but types would have been much better. I didn't convert to TS in this PR because it means turning a lot of other files to typescript as well, and this PR would be even bigger. I'm willing to do that though in a separate PR!
* Add more typescript templates. Currently there is only for react (not counting for angular, which is ts by default). I'm also willing to add templates for the remaining ones.
* Add support for monorepos? Maybe for Lerna based monorepos, as NX for instance has a different way of setting up storybook, using `@nrwl/storybook`.

### CODE REVIEW

File: lib/cli/src/detect.test.js [hunk 1]
```diff
@@ -1,27 +1,288 @@
...
+        dependencies: {
+          rax: '1.0.0',
+        },
+      },
+    },
+  },
+];
+
+describe('Detect', () => {
+  it(`should return type ${projectTypes.HTML} if html option is passed`, () => {
```
Comment (shilman) [inline lib/cli/src/detect.test.js:201]: personal preference: avoid using variables in tests. since tests are documentation, it's common to use the literal constant in the test - more readable / less indirection

File: lib/cli/src/detect.test.js [hunk 2]
```diff
@@ -1,27 +1,288 @@
...
+    it('false if empty devDependency', () => {
+      expect(isStorybookInstalled({ devDependencies: {} }, false)).toBe(false);
+    });
+
+    it('false if no devDependency', () => {
+      expect(isStorybookInstalled({}, false)).toBe(false);
+    });
+
+    supportedFrameworks.forEach((framework) => {
+      it(`true if devDependencies has ${framework} Storybook version`, () => {
```
Comment (shilman) [inline lib/cli/src/detect.test.js:236]: this kind of test would be the exception to the above comment

File: lib/cli/src/detect.test.js [hunk 3]
```diff
@@ -1,27 +1,288 @@
...
+        const devDependencies = {
+          [`@storybook/${framework}`]: '4.0.0-alpha.21',
+        };
+        expect(isStorybookInstalled({ devDependencies }, false)).toBeTruthy();
+      });
+    });
+
+    it('true if forced flag', () => {
```
Comment (shilman) [inline lib/cli/src/detect.test.js:244]: this doesn't have the forced flag and doesn't return true?

File: lib/cli/src/generators/ANGULAR/index.js [hunk 4]
```diff
@@ -14,6 +14,7 @@ import {
   writeFileAsJson,
   copyTemplate,
 } from '../../helpers';
+import { supportedStoryFormats } from '../../project_types';
```
Comment (shilman) [inline lib/cli/src/generators/ANGULAR/index.js:17]: `STORY_FORMAT`?

File: lib/cli/src/helpers.js [hunk 5]
```diff
@@ -316,6 +317,12 @@ export function addToDevDependenciesIfNotPresent(packageJson, name, packageVersi
 export function copyTemplate(templateRoot, storyFormat) {
   const templateDir = path.resolve(templateRoot, `template-${storyFormat}/`);
   if (!fs.existsSync(templateDir)) {
+    // Fallback to CSF plain first, in case format is typescript but template is not available.
+    if (storyFormat === supportedStoryFormats.CSF_TYPESCRIPT) {
```
Comment (shilman) [inline lib/cli/src/helpers.js:321]: WDYT about a warning here? Something like `No TS boilerplate is available for ${framework}, falling back to JS instead.`

File: lib/cli/src/project_types.js [hunk 6]
```diff
@@ -41,6 +50,150 @@ export const supportedFrameworks = [
...
+    },
+  },
+  {
+    preset: projectTypes.REACT_SCRIPTS,
+    // For projects using a custom/forked `react-scripts` package.
+    files: ['/node_modules/.bin/react-scripts'],
+    // For standard CRA projects
+    dependencies: ['react-scripts'],
+    matcherFunction: ({ dependencies, files }) => {
+      return dependencies.every(Boolean) || files.every(Boolean);
```
Comment (shilman) [inline lib/cli/src/project_types.js:122]: COOL 💯

**Review by shilman — APPROVED** [review]
LOOKS great! A few small suggestions

File: lib/cli/src/detect.test.js (hunk 1 above)
Comment (yannbf) [reply lib/cli/src/detect.test.js:201]: You're right, seems much better. Thanks for your comment!

File: lib/cli/src/generators/ANGULAR/index.js (hunk 4 above)
Comment (yannbf) [reply lib/cli/src/generators/ANGULAR/index.js:17]: I thought of defining it like STORY_FORMAT, but to keep consistency I would need to refactor also:
projectTypes
supportedLanguages
supportedFrameworks
supportedTemplates
notInstallableProjectTypes
installableProjectTypes

Which I'm fine with if you think that's the best way to go!

File: lib/cli/src/detect.test.js (hunk 3 above)
Comment (yannbf) [reply lib/cli/src/detect.test.js:244]: Great catch! That was one of the existing tests that I just moved inside a describe. Checking it better, it made no sense, as the method either returns false or a project type. Changed + fixed the other tests which were wrong and added one more!

File: lib/cli/src/helpers.js (hunk 5 above)
Comment (yannbf) [reply lib/cli/src/helpers.js:321]: I thought of doing that, but there are a few problems:

1 - This method is not aware of the framework, unless I do a not so pretty code like `templateRoot.split('/').pop()`, which would be using the template folder, resulting in a message like:
`No TS boilerplate is available for SFC_VUE, falling back to JS instead.`.
It's not great because the folders could have names that are not interesting for the users, like SFC_VUE, or WEBPACK_REACT.
To pass a proper name the method would need to be changed, so that it would receive as parameter from every template generator function.

2 - There is inconsistency in some of the template folders. For instance Angular has `template-csf`, even though it's typescript. As a result, for angular this would be the output:
```
• Detecting project type. ✓      
No TS boilerplate is available, falling back to JS instead. 
• Adding storybook support to your "Angular" app. ✓
```
Leaving the user a bit confused.

I'm up for fixing this, by renaming all folders properly if you want 👍

Comment (shilman) [reply lib/cli/src/helpers.js:321]: Nah, don’t worry about it. Just a suggestion!

File: lib/cli/src/generators/ANGULAR/index.js (hunk 4 above)
Comment (shilman) [reply lib/cli/src/generators/ANGULAR/index.js:17]: I think these string constants should be capitalized. The object constants I’m fine with camel case. But it’s not a big deal to me, which is why I approved the PR as is

Comment (yannbf) [reply lib/cli/src/generators/ANGULAR/index.js]: Cool! Please check the last 4 commits I just pushed

File: lib/cli/src/initiate.js [hunk 7]
```diff
@@ -3,10 +3,11 @@ import chalk from 'chalk';
 import inquirer from 'inquirer';
 import { detect, isStorybookInstalled, detectLanguage } from './detect';
 import { hasYarn } from './has_yarn';
-import types, {
```
Comment (shilman) [inline lib/cli/src/initiate.js]: ❤️

**Comment by shilman** [comment]
@yannbf some test problems:
- storyshots tests running out of memory .. weird (cc @ndelangen )
- e2e tests failing, which looks like maybe a legit failure

Can you look into it, esp the second one?

File: lib/cli/src/project_types.js [hunk 8]
```diff
@@ -195,13 +195,11 @@ export const supportedTemplates = [
...
+  PROJECT_TYPES.ALREADY_HAS_STORYBOOK,
+  PROJECT_TYPES.UPDATE_PACKAGE_ORGANIZATIONS,
 ];
 
-export const installableProjectTypes = Object.values(projectTypes)
+export const installableProjectTypes = Object.values(PROJECT_TYPES)
   .filter((type) => !notInstallableProjectTypes.includes(type))
   .map((type) => type.toLowerCase());
-
-export default projectTypes;
```
Comment (gaetanmaisse) [inline lib/cli/src/project_types.js]: 👍 I'm a big fan of removing `default` export when possible

**Review by gaetanmaisse — APPROVED** [review]

**Comment by gaetanmaisse** [comment]
I triggered a re-run of the 2 failing test workflows and everything is 🟢again on CircleCI. I can't re-run TeamCity jobs as I'm only a guest.

File: lib/cli/src/project_types.js (hunk 8 above)
Comment (yannbf) [reply lib/cli/src/project_types.js]: me too hehe
